docs: add necessary feature gate information for KEP-5073 - Declarative Validation of K8s Native Types With validation-gen

aaron-prindle opened the pull request:
PR related to docs changes necessary for KEP 5073

KEP Issue: https://github.com/kubernetes/enhancements/issues/5073
KEP PR: https://github.com/kubernetes/enhancements/pull/5074


KEP Implementation PRs:
- Migrate to declarative validation - ReplicationController spec.replicas and spec.minReadySeconds fields: https://github.com/kubernetes/kubernetes/pull/130725
- Enable Declarative Validation for ReplicationController: https://github.com/kubernetes/kubernetes/pull/130724
- add declarative validation metrics and associated runtime verification tests: https://github.com/kubernetes/kubernetes/pull/130705
- Add default + optional handling: https://github.com/kubernetes/kubernetes/pull/130706
- add Add CoveredByDeclarative to field error struct: https://github.com/kubernetes/kubernetes/pull/130695
- add feature gates: https://github.com/kubernetes/kubernetes/pull/130478, [Declarative Validation] update Declarative Validation featuregate  kubernetes#130703
- add validation-gen framework: https://github.com/kubernetes/kubernetes/pull/130349


 (related: https://github.com/kubernetes/enhancements/issues/5073#issuecomment-2654441272)

hacktivist123 commented:
Hello @aaron-prindle :wave: please take a look at [Documenting for a release - PR Ready for Review](https://kubernetes.io/docs/contribute/new-content/new-features/#pr-ready-for-review) to get your PR ready for review before Tuesday 25th March 2025 18:00 PDT. Thank you!

aaron-prindle commented:
@hacktivist123 thanks for the information here.  This PR is ready for review and I have changed the status to such.  The docs mention:

> When you complete your content, the documentation person assigned to your feature reviews it. To ensure technical accuracy, the content may also require a technical review from corresponding SIG(s). Use their suggestions to get the content to a release ready state.

I am not sure if our feature has an assigned documentation person, do you happen to know who that might be or if we assign that person?  I'm going to tag @sftim for now to attempt to get into a review queue but happy to change it if there is a more appropriate person for review.  Thanks!

/assign @sftim

tengqm commented on content/en/docs/reference/command-line-tools-reference/feature-gates/DeclarativeValidation.md:
```diff
@@ -0,0 +1,16 @@
+Enables running declarative validation of APIs, where declared. When enabled, APIs with
+declarative validation rules will validate objects using the generated
+declarative validation code and compare the results to the regular imperative validation.
+See DeclarativeValidationTakeover for more.
```
What are the user impacts?
On which component am I supposed to enable this? All of them, including all kubelets?
For developers, what are they supposed to do, regarding the "declarative validation rules"?

aaron-prindle replied:
> What are the user impacts?
For end-users (cluster operators and users deploying applications), there should be no direct impact during the beta phase.  This is more of a drop-in refactor than a new feauture with respect to end-user impact, the goal is for validation to be functionally equivalent.  

> On which component am I supposed to enable this? All of them, including all kubelets?
The kube-apiserver is the only component that this feature gate is enabled/disabled for.  DeclarativeValidation is Beta, default=true and DeclarativeValidationTakeover is default=false.  I have updated the text to make this more clear now, thanks:
`... This feature gate only operates on the kube-apiserver.`

> For developers, what are they supposed to do, regarding the "declarative validation rules"?
For end-users, as mentioned, this should have no impact.  For developers in v1.33 we are still building out the full suite of validators which we will land more of in v1.34.  Developers developing new kubernetes native APIs will be able to use Declarative Validation to develop kubernetes native APIs but this will initially only be relevant to those working on kubernetes native APIs in k/k directly.

tengqm commented on content/en/docs/reference/command-line-tools-reference/feature-gates/DeclarativeValidationTakeover.md:
```diff
@@ -0,0 +1,18 @@
+When disabled, hand-written validation errors are always returned, effectively putting
+declarative validation in a "shadow mode" that monitors but does not affect API responses.
+Note: Although declarative validation aims for functional equivalence with hand-written validation,
+the exact number, format, and content of error messages may differ between the two approaches.
```
Is this an implementation detail or a user facing feature?
Is this gate related to some new metrics?

aaron-prindle replied:
> Is this an implementation detail or a user facing feature?
This is an implementation detail.  This gate is related to a new metric - `declarative_validation_mismatch_total`. 
 I've updated the text here to make the context and metrics more clear:

> When enabled, along with the [DeclarativeValidation](/docs/reference/command-line-tools-reference/feature-gates/DeclarativeValidation.md) feature gate, declarative validation errors are returned directly to the caller, replacing hand-written validation errors for rules that have declarative implementations. When disabled (and `DeclarativeValidation` is enabled), hand-written validation errors are always returned, effectively putting declarative validation in a __mismatch validation mode__ that monitors but does not affect API responses.  This __mismatch validation mode__  allows for the monitoring of the `declarative_validation_mismatch_total` and `declarative_validation_panic_total` metrics which are implementation details for a safer rollout, average user shouldn't need to interact with it directly. Note: Although declarative validation aims for functional equivalence with hand-written validation, the exact number, format, and content of error messages may differ between the two approaches.

sftim commented:
The Prow bot already [suggested](https://github.com/kubernetes/website/pull/49732#event-16287815046) me (and @natalisucks) as reviewers. Happy to take a look.

sftim commented on content/en/docs/reference/command-line-tools-reference/feature-gates/DeclarativeValidation.md:
```diff
@@ -0,0 +1,16 @@
+    defaultValue: true
+    fromVersion: "1.33"
+---
+Enables running declarative validation of APIs, where declared. When enabled, APIs with
```
Is this in-tree APIs or extension APIs from CRDs or aggregated external APIs or some mix of the other.

aaron-prindle replied:
This is for the in-tree APIs that currently have a declarative validation migration/declaration in place (more fields/whole-types will be converted over time in v1.34+).  Currently this is ReplicationSet.[Replias|MinReadySeconds]

https://github.com/kubernetes/kubernetes/pull/130725

I've updated the text to state this now (see below), thanks!

```
Enables declarative validation of in-tree Kubernetes APIs. When enabled, APIs with declarative validation rules (defined using IDL tags in the Go code) will have both the generated declarative validation code and the original hand-written validation code executed.  The results are compared, and any discrepancies are reported via the `declarative_validation_mismatch_total` metric.  The hand-written validation result is returned to the user.
```

sftim commented on content/en/docs/reference/command-line-tools-reference/feature-gates/DeclarativeValidation.md:
```diff
@@ -0,0 +1,16 @@
+Enables running declarative validation of APIs, where declared. When enabled, APIs with
+declarative validation rules will validate objects using the generated
+declarative validation code and compare the results to the regular imperative validation.
+See DeclarativeValidationTakeover for more.
```
> `See DeclarativeValidationTakeover for more.`

Change this to a hyperlink to the documentation (which you, or a colleague from the SIG, will need to write).

aaron-prindle replied:
I am happy to write this documentation but I am currently a bit confused as I was intending for this to essentially just link to the other feature gate documentation that is already a part of this PR.  I have added such a link now:

> The original hand-written validation are still the authoritative validations when this is enabled but this can be changed if the [DeclarativeValidationTakeover feature gate](/docs/reference/command-line-tools-reference/feature-gates/DeclarativeValidationTakeover.md) is enabled in addition to this gate.  ...

Can you clarify what we hope for this additional documentation to have?  Thanks!

sftim commented on content/en/docs/reference/command-line-tools-reference/feature-gates/DeclarativeValidationTakeover.md:
```diff
@@ -0,0 +1,18 @@
+When enabled, declarative validation errors are returned directly to the caller,
+replacing hand-written validation errors for rules that have declarative implementations.
+When disabled, hand-written validation errors are always returned, effectively putting
+declarative validation in a "shadow mode" that monitors but does not affect API responses.
```
```suggestion
declarative validation in a _shadow validation_ mode that monitors but does not affect API responses.
```

sftim commented on content/en/docs/reference/command-line-tools-reference/feature-gates/DeclarativeValidationTakeover.md:
```diff
@@ -0,0 +1,18 @@
+    defaultValue: false
+    fromVersion: "1.33"
+---
+When enabled, declarative validation errors are returned directly to the caller,
```
Change ”declarative validation“ to a link to the docs for this (which you, or a colleague from the SIG, will need to write).

aaron-prindle replied:
I have tentatively pointed this to link to the DeclarativeValidation feature gate file.  I'm happy to write these additional documentation but our current plan was to add such docs in v1.34

sftim reviewed (commented):
I think we should merge this after the docs have merged into the v1.33, so that the feature gate descriptions can link to those docs.

aaron-prindle commented:
> I think we should merge this PR (which adds descriptions for feature gates) after the docs have merged into the v1.33, so that the feature gate descriptions can link to those docs.

> @aaron-prindle do you have plans to write documentation about declarative validation rules? I imagine some will be needed.

There was not plans to write this documentation for the v1.33 release, initially the idea was to only add the feature gate documentation in v1.33.  Declarative Validation is a maintainer+developer facing feature primarily, not an end using facing feature (it unlocks some end users features but not in the near term).  In v1.34 the plan was to add a full set of kubernetes developer facing documentation that explains:
- how to create new Declarative Validation "Validators"/tag (ex: https://github.com/kubernetes/kubernetes/pull/130730)
- how to use Declarative Validation generally and have docs for each "Validator"/tag (will be auto-gen from Docs field here: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/code-generator/cmd/validation-gen/validators/limits.go#L70-L80)
- how to enable/plumb a new k8s type into Declarative Validation with proper gating (ex: https://github.com/kubernetes/kubernetes/pull/130724)
- - how to migrate a field in k/k to use Declarative Validation (ex: https://github.com/kubernetes/kubernetes/pull/130725)

aaron-prindle commented:
/cc @jpbetz

jpbetz commented on content/en/docs/reference/command-line-tools-reference/feature-gates/DeclarativeValidationTakeover.md:
```diff
@@ -0,0 +1,13 @@
+---
+title: DeclarativeValidationMismatchMetric
```
```suggestion
title: DeclarativeValidationTakeover
```

jpbetz commented:
/approve
/lgtm
For technical content

aaron-prindle commented:
@sftim would it be possible to merge the feature gate PR as is for now w/o the additional documentation mentioned?  Currently the plan was for aadditional documentation to land once we have added some additional Validators upstream and then we can document the usage of those additional validators in our initial documentation piece.  We plan to add the documentation noted in the comments above in a future PR

sftim commented:
>  In v1.34 the plan was to add a full set of kubernetes developer facing documentation that explains

We ask for documentation the first time a feature gate is added, unless the feature gate doesn't have an effect (in which case, usually the project should revert the feature gate in the code).

If someone is writing an extension API server, and they want to use this, what do they need to know? We should aim to document a minimum useful level of detail.

jpbetz commented:
> >  In v1.34 the plan was to add a full set of kubernetes developer facing documentation that explains
> 
> 
> 
> We ask for documentation the first time a feature gate is added, unless the feature gate doesn't have an effect (in which case, usually the project should revert the feature gate in the code).
> 
> If someone is writing an extension API server, and they want to use this, what do they need to know? We should aim to document a minimum useful level of detail.

This one is tricky. The feature gate switches the implementation of validation (for only 2 fields in ReplicationController in 1.33).  When the feature gates are on there is an additional metrics.

The apiserver, as a library, cannot reasonably use it yet because the validation tags implemented are only sufficient for the fields we've migrated so far.

I agree with the sentiment that we should document features as we go.  I'm struggling a bit to find a somewhere we could write something meaningful.  We're adding `validation-gen` next to `defaulter-gen` and others.  Most of the documentation for the adjacent code exists in the readme in package docs and readme files. https://kubernetes.io/docs/tasks/extend-kubernetes/setup-extension-api-server/ does not feel like a good place to put a page.

I was expecting us to do the bulk of declarative validation in the code and in the contrib repository (for guides on how to migrate).

Should we introduce a new page for "apiserver development" or "extension apiserver internals"?  Does that fit into the website?

sftim commented:
I agree that this doesn't seem to need much documentation, but I think it needs some.

> Should we introduce a new page for "apiserver development" or "extension apiserver internals"? Does that fit into the website?

Maybe just a stub. We already have https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/apiserver-aggregation/ and that page could link to the stub.
[Ideally: we eventually document more than a stub, but that's a separate improvement]

We don't need any documentation aimed at people contributing to Kubernetes itself (but if we did, it would live within https://k8s.dev/docs/ anyway and not on this site).

- When this change is stable, what differences will an end-user cluster administrator see?
- During the journey to stability, what do people need to know about declarative validation around cluster upgrades?
- Who might need to turn off this feature gate (v1.33 and also our expectations for future versions) and why.

aaron-prindle commented:
> Maybe just a stub. We already have https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/apiserver-aggregation/ and that page could link to the stub.
> [Ideally: we eventually document more than a stub, but that's a separate improvement]
> 
> We don't need any documentation aimed at people contributing to Kubernetes itself (but if we did, it would live within https://k8s.dev/docs/ anyway and not on this site).
> 
> When this change is stable, what differences will an end-user cluster administrator see?
> During the journey to stability, what do people need to know about declarative validation around cluster upgrades?
> Who might need to turn off this feature gate (v1.33 and also our expectations for future versions) and why.

@sftim  - I have submitted a PR to add documentation that addresses the questions in the quoted comment and links this documentation in the suggested page.  Link to the PR is here: https://github.com/kubernetes/website/pull/50265

aaron-prindle commented:
@sftm I have addressed all feedback from the initial round of reviews and the related PR for a separate Declarative Validation docs page spawned from comments here (https://github.com/kubernetes/website/pull/50265) is merged now.  I believe this is ready for another round of reviews now, thanks!

rayandas commented:
Hello @aaron-prindle :wave:! I'm reaching out from the Docs team. Just checking in as we approach **Docs Freeze** on **8th April, 2025 18:00 PDT**.
This documentation appears to still be under review. To meet the Docs Freeze, this PR must have a technical review as well as `lgtm` and `approve` labels applied, without any unaddressed comments or concerns from SIG Docs.
The status of this enhancement is marked as `At risk for docs freeze`. Thank you!

tengqm commented:
/approve
